### ISSUE TITLE
Source replacements are ignored when producing compilation units

### ISSUE THREAD

**Issue by P-E-Meunier** [rust-lang/cargo#14821]
### Problem

When replacing sources in `.cargo/config.toml` so that two different sources point to the same vendored directory, cargo identifies a single package from two different sources as different, giving rise to a confusing error message where a type from a file is claimed to be distinct from itself:

```
error[E0308]: mismatched types
 --> src/main.rs:2:10
  |
2 |     b::b(c::C {});
  |     ---- ^^^^^^^ expected `c::C`, found `C`
  |     |
  |     arguments to this function are incorrect
  |
  = note: `C` and `c::C` have similar names, but are actually distinct types
note: `C` is defined in crate `c`
 --> /tmp/cargo-vendor-test/a/vendor/c/src/lib.rs:1:1
  |
1 | pub struct C {}
  | ^^^^^^^^^^^^
note: `c::C` is defined in crate `c`
 --> /tmp/cargo-vendor-test/a/vendor/c/src/lib.rs:1:1
  |
1 | pub struct C {}
  | ^^^^^^^^^^^^
  = note: perhaps two different versions of crate `c` are being used?
note: function defined here
 --> /tmp/cargo-vendor-test/a/vendor/b/src/lib.rs:1:8
  |
1 | pub fn b(_c: c::C) {}
  |        ^
```

### Steps

1. Clone https://github.com/P-E-Meunier/cargo-vendor-test
2. Run `cargo vendor`, which yields a wrong output (as per #14729).
3. The correct output should be:
```
[source."sparse+http://localhost:8080/registry-y/"]
registry = "sparse+http://localhost:8080/registry-y/"
replace-with = "vendored-sources"

[source."sparse+http://localhost:8080/registry-x/"]
registry = "sparse+http://localhost:8080/registry-x/"
replace-with = "vendored-sources"

[source.vendored-sources]
directory = "vendor"
```
4. Run `cargo build`

### Version

```text

```

**Comment by weihanglo** [comment]
Ran `cargo vendor` didn't give me that output. With which version of Cargo under which directory should I run `cargo vendor`?

**Comment by P-E-Meunier** [comment]
So, in my example repo, you need to run those steps from subdirectory `a`.
In #14729, you additionally need to start a fake private registry before using the following command from the root of the repo:
```
python -m http.server 8080 --bind 127.0.0.1 --directory .
```

**Comment by weihanglo** [comment]
> In https://github.com/rust-lang/cargo/issues/14729, you additionally need to start a fake private registry before

You need to start the fake registry in either of those, otherwise Cargo cannot connect to them.

Anyway I can reproduce it now. This reminds me https://github.com/rust-lang/cargo/issues/10310 though they don't seem to be the same, and also one of the two `c` crate is vendored. The issue is kind of hitting several bad cases of `cargo vendor`?

**Comment by P-E-Meunier** [comment]
It is indeed. This isn't an artificial case though, it was triggered by migrating a registry from one domain to another: some versions of some crates are mirrored on both, leading to this situation.

Fortunately, the fix for #14729 doesn't have any unintended effects: the only thing it does is replace more sources in `.cargo/config.toml`, which can't hurt if these registries aren't used, and fixes the issue when they are.

The fix for this issue is a little different, since it changes the rustc invocations when sources are replaced:
- Currently, Cargo produces "metadata" or "extra filenames" to be prepended to build products based on the *original* source of the dependency. If you replaced a source with a local one in `.cargo/config.toml`, the metadata doesn't change, meaning that Cargo considers the two sources (original and replaced) to be the same. IIUC, both would be valid since an original crate and its replacement never appear in the same `rustc` invocation.
- All #14822 does is to use the replaced source to produce the metadata instead, meaning that (1) a replaced source won't be considered the same as its original source but (2) two sources replaced to the same one will be considered equal.

**Comment by epage** [comment]
For me, it seems suspect to try to take items from different sources and claim they are from the same source just because they are vendored in the same location.

**Comment by P-E-Meunier** [comment]
@epage Why? When you're compiling a vendored dependency, the compilation isn't strictly the same as compiling from `~/.cargo/cache/…`. And indeed the difference can be significant when using private registries, to the point of preventing the compilation altogether when vendoring, while the compilation from upstream sources works fine.

**Comment by epage** [comment]
> And indeed the difference can be significant when using private registries, to the point of preventing the compilation altogether when vendoring, while the compilation from upstream sources works fine.

Could you help me understand why compilation fails when vendoring dependencies when the build works without vendoring?

**Comment by P-E-Meunier** [comment]
TLDR; source replacement (`.cargo/config.toml`) works differently with and without vendoring.

I gave a reproduction illustrating this in the first message of this issue:

- In the original version, the same lib (`c`) from two different registries is/are replaced with `c` from one of the registries, and Cargo is correctly able to tell that they are the same.
- In the vendored version, the source replacement are ignored, and the sources are vendored as two different crates with the same name, whereas they should be one (since a source replacement is involved).

**Comment by epage** [comment]
So in your situation, for `Cargo.toml`, you have:
- a depends on
  - y's b
  - y's c
- b depends on 
  - x's c
- c

For the registry you have
- y's b depends on
  - y's c
- y's c
- x's c

b's index entry: https://github.com/P-E-Meunier/cargo-vendor-test/blob/main/registry-y/1/b

index schema: https://doc.rust-lang.org/cargo/reference/registry-index.html#json-schema

So in this case, the registry is reporting information that does not match the `Cargo.toml` within the `.crate` file (presuming the `.crate` was generated from the source in that tree) and Cargo is not identifying this discrepancy and reporting it to the user in some way, instead merrily going on.

To me, the Index is malformed and having `cargo vendor` mimic a malformed Index's behavior does not seem "correct".

**Comment by P-E-Meunier** [comment]
What is "malformed" about that index?

The main issue here is that `cargo vendor` doesn't output source replacements for some of the sources in that particular case. Maybe Cargo can already do this without https://github.com/rust-lang/cargo/pull/14822, but I'm not sure how.

**Comment by epage** [comment]
> What is "malformed" about that index?

There is a discrepancy in how `b` is defined
- `Cargo.toml`: c's registry is `x`
- Index: c's registry is default (i.e. `y`)
